ovgs.proto

```diff
@@ -1,175 +1,241 @@
-    // readDomainCert reveals the details of the certificate.
-    rpc readDomainCert(readDomainCertRequest) returns (readDomainCertResponse);
-    // deleteDomainCert deletes the cert from the database.
-    rpc deleteDomainCert(deleteDomainCertRequest) returns (deleteDomainCertResponse);
-    // ownershipVoucher issues ownership voucher and returns TPM public key for the serial number
-    rpc ownershipVoucher(ownershipVoucherRequest) returns (ownershipVoucherResponse);
+  // CreateGroup creates a group as a child of an existing group.
+  // Errors will be returned:
+  // ALREADY_EXISTS if group already exists
+  // PERMISSION_DENIED if the user doesn't have access to the parent group
+  rpc CreateGroup(CreateGroupRequest) returns (CreateGroupResponse);
+  // DeleteGroup deletes a named group.
+  // Errors will be returned:
+  // NOT_FOUND if the group doesn't exist
+  // FAILED_PRECONDITION if the parent is set but does not exist
+  // PERMISSION_DENIED if user doesn't have access to parent group
+  rpc DeleteGroup(DeleteGroupRequest) returns (DeleteGroupResponse);
+  // GetGroup returns the domain-certs (keyed by id), serials,
+  // and user/role mappings for that group and all the children.
+  // Errors will be returned:
+  // NOT_FOUND if the group doesn't exist
+  // PERMISSION_DENIED if the user doesn't have access to the group
+  rpc GetGroup(GetGroupRequest) returns (GetGroupResponse);
+
+  // CreateUserRole assigns a role to a user in a named group.
+  // Errors will be returned:
+  // ALREADY_EXISTS if user already exists in the group
+  // PERMISSION_DENIED if the user doesn't have access to the group
+  rpc CreateUserRole(AssignUserRoleRequest) returns (AssignUserRoleResponse);
+  // DeleteUserRole removes roles of a user in a named group.
+  // Errors will be returned:
+  // NOT_FOUND if the group doesn't exist
+  // PERMISSION_DENIED if user doesn't have access to the group
+  rpc DeleteUserRole(AssignUserRoleRequest) returns (AssignUserRoleResponse);
+  // GetUserRole returns the roles the user is assigned.
+  // Errors will be returned:
+  // NOT_FOUND if the group doesn't exist
+  // PERMISSION_DENIED if the user doesn't have access to the group
+  rpc GetUserRole(GetUserRoleRequest) returns (GetUserRoleResponse);
+
+  // CreateSerial assigns the serial to that group.
+  // Errors will be returned:
+  // ALREADY_EXISTS if serial_number already exists in the group
+  // PERMISSION_DENIED if the user doesn't have access to the group
+  rpc CreateSerial(CreateSerialRequest) returns (CreateSerialResponse);
+  // DeleteSerial removes the serial from the group.
+  // Errors will be returned:
+  // NOT_FOUND if the group doesn't exist
+  // PERMISSION_DENIED if user doesn't have access to the group
+  rpc DeleteSerial(DeleteSerialRequest) returns (DeleteSerialResponse);
+  // GetSerial returns all the facts about the serial number.
+  // Errors will be returned:
+  // NOT_FOUND if the group doesn't exist
```

Comment:
Nit: NOT_FOUND if the serial doesn't exist

Reply:
GetSerialRequest only takes the serial_number as an input parameter and returns the groups the serial shows up in.

You can't delete a group without removing all serials assigned to it, which means any group returned in the response must exist.

Is there something else you mean by 'group doesn't exist'?